src/main/java/updater/DownloadFile.java

```diff
@@ -0,0 +1,106 @@
+package updater;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.*;
+import java.net.URI;
+import java.net.URLConnection;
+import java.util.function.Consumer;
+
+/**
+ * A class in charge of downloading file.
+ */
+public class DownloadFile {
+    private static final Logger logger = LogManager.getLogger(DownloadFile.class.getName());
+
+    public static final int CONNECTION_TIMEOUT = 15000;
+    public static final int READ_CONNECTION_TIMEOUT = 30000;
+
+    private final URI source;
+    private final File dest;
+    private long totalFileSizeInBytes;
+
+    private final Consumer<Double> onDownload;
+
+    /**
+     * @param source URI of file source, expect to be a valid URL
+     * @param dest local destination file
+     * @param onDownload callback that listen to download stream
+     */
+    public DownloadFile(URI source, File dest, Consumer<Double> onDownload) {
+        this.source = source;
+        this.dest = dest;
+        this.onDownload = onDownload;
+    }
+
+    /**
+     * Start download from source to destination as specified in constructor
+     *
+     * @return true if download is successful, else false
+     */
+    public boolean startDownload() {
+        URLConnection sourceConnection = null;
+
+        try {
+            sourceConnection = setupDownloadConnection(source);
+        } catch (IOException e) {
+            logger.error("URI of source file is not a well-formed URL");
+            return false;
+        }
+
+        assert sourceConnection != null;
+
+        totalFileSizeInBytes = sourceConnection.getContentLengthLong();
+
+        try (
+                InputStream inputStream = setupStreamFromSource(sourceConnection);
+                OutputStream outputStream = setupStreamToDest(dest);
+        ) {
+            if (download(inputStream, outputStream) != totalFileSizeInBytes) {
```

Comment:
Hmmm, the thing is I can't reproduce the error I am trying to capture here manually. What I saw last time was IOUtils printed a message for socket time out (I think Github server closed the connection), but it was not caught. In this case, the only way to check if download ended successfully is by checking file size. I agree the case does not seems possible, but I have yet to encounter the same error.